internal/backup/volumesnapshot_action.go

```diff
@@ -144,18 +154,98 @@ func (p *VolumeSnapshotBackupItemAction) Execute(item runtime.Unstructured, back
+	for k, v := range annotations {
+		pb += fmt.Sprintf("\"%s\":\"%s\",", k, v)
+	}
+	pb = strings.Trim(pb, ",")
+	pb += "}}}"
+	if _, err := snapshotClient.SnapshotV1().VolumeSnapshots(vs.Namespace).Patch(context.TODO(),
+		vs.Name, types.MergePatchType, []byte(pb), metav1.PatchOptions{}); err != nil {
+		p.Log.Errorf("Fail to patch volumesnapshot with content %s: %s.", pb, err.Error())
+		return nil, nil, "", nil, errors.WithStack(err)
+	}
+
+	annotations[util.MustIncludeAdditionalItemAnnotation] = "true"
 	// save newly applied annotations into the backed-up volumesnapshot item
 	util.AddAnnotations(&vs.ObjectMeta, annotations)
 
 	vsMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&vs)
 	if err != nil {
-		return nil, nil, errors.WithStack(err)
+		return nil, nil, "", nil, errors.WithStack(err)
 	}
 
 	p.Log.Infof("Returning from VolumeSnapshotBackupItemAction with %d additionalItems to backup", len(additionalItems))
 	for _, ai := range additionalItems {
 		p.Log.Debugf("%s: %s", ai.GroupResource.String(), ai.Name)
 	}
 
-	return &unstructured.Unstructured{Object: vsMap}, additionalItems, nil
+	// The operationID is of the form <namespace>/<volumesnapshot-name>/<started-time>
+	operationID := vs.Namespace + "/" + vs.Name + "/" + time.Now().Format(time.RFC3339)
+	itemToUpdate := []velero.ResourceIdentifier{
+		{
+			GroupResource: kuberesource.VolumeSnapshots,
+			Namespace:     vs.Namespace,
+			Name:          vs.Name,
+		},
+	}
+
+	return &unstructured.Unstructured{Object: vsMap}, additionalItems, operationID, itemToUpdate, nil
+}
+
+func (p *VolumeSnapshotBackupItemAction) Name() string {
+	return "VolumeSnapshotBackupItemAction"
+}
+
+func (p *VolumeSnapshotBackupItemAction) Progress(operationID string, backup *velerov1api.Backup) (velero.OperationProgress, error) {
+	progress := velero.OperationProgress{}
+	if operationID == "" {
+		return progress, biav2.InvalidOperationIDError(operationID)
+	}
+	// The operationID is of the form <namespace>/<volumesnapshot-name>/<started-time>
+	operationIDParts := strings.Split(operationID, "/")
+	if len(operationIDParts) != 3 {
+		p.Log.Errorf("invalid operation ID %s", operationID)
+		return progress, biav2.InvalidOperationIDError(operationID)
+	}
+	var err error
+	if progress.Started, err = time.Parse(time.RFC3339, operationIDParts[2]); err != nil {
+		p.Log.Errorf("error parsing operation ID's StartedTime part into time %s: %s", operationID, err.Error())
+		return progress, errors.WithStack(err)
+	}
+
+	_, snapshotClient, err := util.GetClients()
+	if err != nil {
+		return progress, errors.WithStack(err)
+	}
+
+	vs, err := snapshotClient.SnapshotV1().VolumeSnapshots(operationIDParts[0]).Get(context.Background(), operationIDParts[1], metav1.GetOptions{})
+	if err != nil {
+		p.Log.Errorf("error getting volumesnapshot %s/%s: %s", operationIDParts[0], operationIDParts[1], err.Error())
+		return progress, errors.WithStack(err)
+	}
+
+	if vs.Status == nil {
+		return progress, nil
```

Comment:
maybe add a debug log here that VS is yet to have a status or something

Reply:
Done